Document/0x05b-Basic-Security_Testing.md

```diff
@@ -302,15 +337,112 @@ uid=0(root) gid=0(root) groups=0(root),1004(input),1007(log),1011(adb),1015(sdca
+###### From Rooted Devices
+TODO via adb pull
+###### From Non-Rooted Devices
+
+This works on both rooted and non-rooted devices:
+
+```
+$ adb shell pm path sg.vp.owasp_mobile.omtg_android
+package:/data/app/sg.vp.owasp_mobile.omtg_android-Q0Y29-poHT2w8912zTxXNw==/base.apk
+
+$ adb pull /data/app/sg.vp.owasp_mobile.omtg_android-Q0Y29-poHT2w8912zTxXNw==/base.apk
+/data/app/sg.vp.owasp_mobile.omtg_android-Q0Y29-poHT2w8912zTxXNw==/base.apk: 1 file pulled. 17.9 MB/s (5801767 bytes in 0.309s)
+```
+
+#### Installing Apps
+
+Use `adb install` to install an APK on an emulator or connected device.
+
+```bash
+adb install path_to_apk
+```
+
+Note that if you have the original source code and use Android Studio, you do not need to do this because Android Studio handles the packaging and installation of the app for you.
+
+#### Information Gathering
+##### Installed Apps
+TODO
+```shell
+$ adb shell pm list packages
+package:com.android.sdm.plugins.connmo
+package:com.google.android.googlequicksearchbox
+package:com.android.providers.calendar
+package:com.android.providers.media
+package:com.google.android.onetimeinitializer
+package:com.google.android.ext.shared
+package:com.android.wallpapercropper
+package:sg.vp.owasp_mobile.omtg_android
+```
+
+
+
+##### App Basic Information
+TODO aapt, dumpsys
+###### Sandbox
+
+```shell
+$ adb shell pm path sg.vp.owasp_mobile.omtg_android
+package:/data/app/sg.vp.owasp_mobile.omtg_android-Q0Y29-poHT2w8912zTxXNw==/base.apk
+```
+
+```shell
+$ adb shell ls /data/data/sg.vp.owasp_mobile.omtg_android
+```
+###### Permissions
+TODO from androidmanifest or aapt dump
+
+```
+aapt dump permissions omtg.apk
+```
+###### Native Libs
+TODO APK/libs or from device; or frida / objection when running
+
+```
+bullhead:/ # ls /data/data/sg.vp.owasp_mobile.omtg_android/lib
+libdatabase_sqlcipher.so libnative.so libsqlcipher_android.so libstlport_shared.so
+```
+
+Using objection it is also straightforward as you already land in the sandbox and can directly use `cd` and `ls` commands to look around:
+
+```
+$ objection -g sg.vp.owasp_mobile.omtg_android explore
+
+...g.vp.owasp_mobile.omtg_android on (google: 8.1.0) [usb] # ls lib                                                                                                           
+Type    Last Modified            Read    Write    Hidden    Size       Name
+------  -----------------------  ------  -------  --------  ---------  ------------------------
+File    2018-06-06 08:16:34 GMT  True    False    False     13.4 KiB   libnative.so
+File    2018-06-06 08:16:34 GMT  True    False    False     181.9 KiB  libdatabase_sqlcipher.so
+File    2018-06-06 08:16:34 GMT  True    False    False     357.3 KiB  libstlport_shared.so
+File    2018-06-06 08:16:34 GMT  True    False    False     2.2 MiB    libsqlcipher_android.so
+```
+
+###### ...
+
+
+##### Accessing App Data (what to expect, where, sandbox structure)
+TODO /data/data
+##### Monitoring System Logs
```

Comment:
This also described in the test case where we check for sensitive information in logs. Would make sense to also describe it here in general, but just need to check that we have no redundant information. Maybe need to shift some content from the test case here.

Reply:
I think that's a good idea, having the basic testing techniques here and actually using them on the test cases. Same goes for the previous two comments above.

For example, I'd think that obtaining, inspecting the AndroidManifest (or the main Info.plist on iOS) and knowing what's inside can be consider a basic technique. We could have it here and refer from the test cases. 

What if we build this "Information Gathering" section on a way that
- it explains how to access fundamental info like
  - installed apps
  - basic info about the app like what's in the APK vs sandbox, all relevant paths (all of this we could easily get with e.g. objection env and ls commands; also very useful on iOS!)
  - what's the manifest (or Info.plist) and what it contains
- but then for the details (e.g. permissions) and what to test/verify, we refer to the specific chapter?

Reply:
please let me know what you think and I will modify the corresponding TODOs accordingly.

Reply:
We should mention and explain the basic commands for info gathering here and use it then accordingly in the context of the test cases to get the expected outcome. That would make sense, as some of these techniques are important to know and are the basis for mobile testing. We just need to be careful of not referencing back all the time from the test case back to this chapter as otherwise the flow of reading in the test case is disrupted. 
We also need to be careful of not explaining something that is already there in 0x05a and 0x06a (for example the "what's the manifest" you mentioned is explained in 0x05a and also mentioned in many test cases of the android chapter). 

@TheDauntless What is your opinion on this?

Reply:
My ideal situation would be that you have the describing sections on Android ("What is the manifest, what are dangerous permissions, where does an application store data, ...") in separate files, and the testing section would follow the MASVS per control and explain how that specific control can be tested.

For example:
## 6.1 - The app only requests the minimum set of permissions necessary.
Extract the Android Manifest as explained in [Information gathering - Extracting the manifest](#). For each of the permissions, verify if they are needed for the application to function. 

The follow tools can automate this process:
- MobSF - Shows full overview of requested permissions
- AwesomeTool - Maps permissions to functionality with screenshots
----------
But that's probably too much change right now... I don't feel that "referencing back all the time" is an issue. If you are new, and you read through the test cases, you will end up at "Open the manifest as explained in [Information gathering - Extracting the manifest](#)". The first time, you will click that link, the second time, you will remember what you learned earlier (and think "Great, I know how to do that already!").

So:
* **General section**: Explain what a manifest is, why it is needed, what it does (developer point of view)
* **Basic testing**: Explain how to acquire it, which difficulties you can face (e.g. obfuscated) (security audit point of view / blackbox)
* **Test case**: "Search for defined permissions and map them to application functionality", "Verify defined permissions' protection levels", ...

Reply:
Agree, this makes sense. So we stick to our structure we have: 
- General section would be 0x05a/0x06a
- Basic testing would be 0x05b/0x06b
and then the individual test cases.
Just need to align and structure iOS/Android chapters properly, so it has a proper flow and there is no redundancy. 
I like the points you mentioned (developer point of view/security audit point of view etc.). We are somehow doing this already, but need to stick more to this.